walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java

```diff
@@ -15,9 +15,119 @@
 package com.google.sps;
 
 import java.util.Collection;
+import java.util.Collections;
+import java.util.Set;
+import java.util.ArrayList;
+import java.util.Arrays;
 
 public final class FindMeetingQuery {
   public Collection<TimeRange> query(Collection<Event> events, MeetingRequest request) {
-    throw new UnsupportedOperationException("TODO: Implement this method.");
+    if (request.getDuration() <= TimeRange.WHOLE_DAY.duration()){
+        ArrayList<TimeRange> mandatoryTimes = findAvailableMeetingTimes(events, request, true);
+        ArrayList<TimeRange> optionalTimes =  findAvailableMeetingTimes(events,request, false);
+        Collection<String> mandatoryAttendees = request.getAttendees();
+        Collection<String> optionalAttendees = request.getOptionalAttendees();
+
+        if(optionalAttendees.size()== 0){
+            return mandatoryTimes;
+        }
+        if(mandatoryAttendees.size()== 0){
+            return optionalTimes;
+        }
+        ArrayList<TimeRange> commonTimes = findCommonTimes(mandatoryTimes,optionalTimes, request);
+        if(commonTimes.size() == 0){
+            return mandatoryTimes;
+        }
+        else{
+            return commonTimes;
+        }
+    
+    }
+    else {
+      return Collections.emptyList();
+    }
+  }
+  
+  //for isMandatoryMeeting, true means mandatory attendees and false means optional attendees
+   public ArrayList<TimeRange> findAvailableMeetingTimes(Collection<Event> events, MeetingRequest request, boolean isMandatoryMeeting){
```

Comment:
I'd probably name the bool parameter something more like `findTimesForMandatoryAttendees` or invert the condition and call it `includeOptionalAttendees`, since it controls whether optional attendees are included in the search, not whether the meeting itself is mandatory.

Reply:
done